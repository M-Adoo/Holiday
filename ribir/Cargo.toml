[package]
authors.workspace = true
categories.workspace = true
description.workspace = true
documentation.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "ribir"
readme.workspace = true
repository = "https://github.com/RibirX/Ribir/ribir"
version.workspace = true

[dependencies]
lyon_algorithms.workspace = true
lyon_path.workspace = true
ribir_core = {path = "../core", version = "0.0.0", features = ["png"]}
ribir_gpu = {path = "../gpu", version = "0.0.0"}
ribir_widgets = {path = "../widgets", version = "0.0.0"}

[dev-dependencies]
env_logger.workspace = true
png.workspace = true

[features]
default = ["wgpu_gl"]
wgpu_gl = ["ribir_gpu/wgpu_gl"]
