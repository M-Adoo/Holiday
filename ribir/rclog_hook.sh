#!/bin/bash

if [ "$MERGE_CHANGELOG" = true ]; then
  echo "$(rclog -t $NEW_VERSION -p ./CHANGELOG.md merge)" >| ./CHANGELOG.md
fi
