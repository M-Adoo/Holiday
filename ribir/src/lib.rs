pub use ribir_core as core;
pub use ribir_widgets as widgets;
pub mod prelude {
  pub use ribir_core::prelude::*;
  pub use ribir_widgets::prelude::*;
  pub mod app {

    #[cfg(feature = "wgpu_gl")]
    pub fn run(root: super::Widget) {
      let mut app = super::Application::default();
      let wnd = app.new_window(|native_wnd, ctx| {
        let size = native_wnd.inner_size();
        let p_backend = super::AppContext::wait_future(gpu::wgpu_backend_with_wnd(
          &native_wnd,
          super::DeviceSize::new(size.width, size.height),
          None,
          None,
          ctx.shaper.clone(),
        ));
        super::Window::new(native_wnd, p_backend, root.into_widget(), ctx)
      });
      app.exec(wnd);
    }
  }
}

use prelude::*;
#[cfg(feature = "wgpu_gl")]
pub fn wgpu_headless_wnd(root: Widget, ctx: AppContext, size: DeviceSize) -> Window {
  let p_backend = AppContext::wait_future(gpu::wgpu_backend_headless(
    size,
    None,
    None,
    ctx.shaper.clone(),
  ));
  Window::new(
    ribir_core::window::MockRawWindow {
      size: Size::from_untyped(size.to_f32().to_untyped()),
      ..Default::default()
    },
    p_backend,
    root,
    ctx,
  )
}
