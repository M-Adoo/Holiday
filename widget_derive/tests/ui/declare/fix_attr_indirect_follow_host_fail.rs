use ribir::prelude::*;

fn fix_attr_indirect_follow_host_widget_pass(ctx: &mut BuildCtx) {
  declare! {
    SizedBox {
      id: a,
      size: Size::zero(),
      cursor: b.cursor,
      SizedBox {
        id: b,
        size: Size::zero(),
        cursor: if a.size.area() > 0 {
          CursorIcon::Hand
        } else {
          CursorIcon::RightArrow
        } ,
      }
    }
  };
}

fn fix_attr_indirect_follow_host_attr_fail(ctx: &mut BuildCtx) {
  declare! {
    SizedBox {
      id: a,
      size: Size::zero(),
      cursor: b.cursor,
      SizedBox {
        id: b,
        size: Size::zero(),
        cursor: a.cursor
      }
    }
  };
}

fn main() {}
