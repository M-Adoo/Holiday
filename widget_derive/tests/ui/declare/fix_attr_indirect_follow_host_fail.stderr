error: Can't init widget because circle follow: a.cursor ～> b , b.cursor ～> a
  --> tests/ui/declare/fix_attr_indirect_follow_host_fail.rs:6:11
   |
6  |       id: a,
   |           ^
7  |       size: Size::zero(),
8  |       cursor: b.cursor,
   |       ^^^^^^  ^
9  |       SizedBox {
10 |         id: b,
   |             ^
11 |         size: Size::zero(),
12 |         cursor: if a.size.area() > 0 {
   |         ^^^^^^     ^
   |
note: If the circular is your want and will finally not infinite change,you should break the init circle and declare some follow relationship in `data_flow`, and remember use `#[skip_nc]` attribute to skip the no change trigger of the field modifyto ignore infinite state change trigger.
  --> tests/ui/declare/fix_attr_indirect_follow_host_fail.rs:8:7
   |
8  |       cursor: b.cursor,
   |       ^^^^^^
...
12 |         cursor: if a.size.area() > 0 {
   |         ^^^^^^

error: Can't init widget because circle follow: a.cursor ～> b , b.cursor ～> a
  --> tests/ui/declare/fix_attr_indirect_follow_host_fail.rs:25:11
   |
25 |       id: a,
   |           ^
26 |       size: Size::zero(),
27 |       cursor: b.cursor,
   |       ^^^^^^  ^
28 |       SizedBox {
29 |         id: b,
   |             ^
30 |         size: Size::zero(),
31 |         cursor: a.cursor
   |         ^^^^^^  ^
   |
note: If the circular is your want and will finally not infinite change,you should break the init circle and declare some follow relationship in `data_flow`, and remember use `#[skip_nc]` attribute to skip the no change trigger of the field modifyto ignore infinite state change trigger.
  --> tests/ui/declare/fix_attr_indirect_follow_host_fail.rs:27:7
   |
27 |       cursor: b.cursor,
   |       ^^^^^^
...
31 |         cursor: a.cursor
   |         ^^^^^^
